crates/witness-generator/src/rpc_generator.rs

```diff
@@ -417,3 +424,163 @@ impl TryFrom<RpcFlatHeaderKeyValues> for HeaderMap {
+
+        // Generate to path
+        let target_dir = tempfile::tempdir()
+            .expect("Failed to create temporary directory for blocks and witnesses");
+        rpc_bw
+            .generate_to_path(target_dir.path())
+            .await
+            .expect("Failed to generate blocks and witnesses to path");
+
+        assert_eq!(
+            std::fs::read_dir(target_dir.path())
+                .expect("Failed to read directory")
+                .count(),
+            2,
+            "Expected 2 files in temporary directory"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_concrete_block_num() {
+        if std::env::var("RPC_URL").is_err() {
+            eprintln!("skipping test: set RPC_URL to run this test");
+            return;
+        }
+
+        let latest_block_number = EthApiClient::<
+            TransactionRequest,
+            Transaction,
+            Block,
+            Receipt,
+            Header,
+        >::block_by_number(
+            &build_base_rpc().build().unwrap().client,
+            BlockNumberOrTag::Latest,
+            false,
+        )
+        .await
+        .expect("Failed to fetch latest block")
+        .unwrap()
+        .header
+        .number;
+
+        // Fetch a non-tip block number
+        let block_number = latest_block_number - 1;
+
+        let rpc_bw = build_base_rpc()
+            .block(block_number)
+            .build()
+            .expect("Failed to build RPC Blocks and Witnesses");
+
+        // Generate to Vector
+        let bws = rpc_bw
+            .generate()
+            .await
+            .expect("Failed to generate blocks and witnesses");
+
+        assert_eq!(bws.len(), 1, "Expected 1 block and witness");
+        assert_eq!(
+            bws[0].block_and_witness.block.number, block_number,
+            "Expected block number to match"
+        );
+
+        // Generate to path
+        let target_dir = tempfile::tempdir()
+            .expect("Failed to create temporary directory for blocks and witnesses");
+        rpc_bw
+            .generate_to_path(target_dir.path())
+            .await
+            .expect("Failed to generate blocks and witnesses to path");
+
+        assert_eq!(
+            std::fs::read_dir(target_dir.path())
+                .expect("Failed to read directory")
+                .count(),
+            1,
+            "Expected 1 files in temporary directory"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_live_blocks() {
+        if std::env::var("RPC_URL").is_err() {
+            eprintln!("skipping test: set RPC_URL to run this test");
+            return;
+        }
+
+        let stop_token = CancellationToken::new();
+
+        // Spawn a task to cancel the token after ~12s which should be enough time to fetch at least one block.
+        tokio::spawn({
+            let stop_token = stop_token.clone();
+            async move {
+                tokio::time::sleep(std::time::Duration::from_secs(12)).await;
```

Comment:
I was tempted to mark this "slow", but for now leaning on the safe side and always running it. Another strategy is avoid the 12s wait and inspect for a file appearing and stop as soon as possible, but still not sure that changes thing much (maybe safe 6s on avg). Ended up on the simpler side.